pkg/candevice/device_linux.go

```diff
@@ -0,0 +1,425 @@
+	c, err := netlink.Dial(unix.NETLINK_ROUTE, &netlink.Config{})
+	if err != nil {
+		return fmt.Errorf("couldn't dial netlink socket: %w", err)
+	}
+	defer c.Close()
+
+	ifi := &ifInfoMsg{}
+	ifi.Index = d.index
+	req, err := d.newRequest(unix.RTM_NEWLINK, ifi)
+	if err != nil {
+		return fmt.Errorf("couldn't create netlink request: %w", err)
+	}
+
+	li := &linkInfoMsg{}
+	li.linkType = canLinkType
+	li.info.BitTiming.Bitrate = bitrate
+	ae := netlink.NewAttributeEncoder()
+	ae.Nested(unix.IFLA_LINKINFO, li.encode)
+	liData, err := ae.Encode()
+	if err != nil {
+		return fmt.Errorf("couldn't encode message: %w", err)
+	}
+	req.Data = append(req.Data, liData...)
+
+	res, err := c.Execute(req)
+	if err != nil {
+		return fmt.Errorf("couldn't set bitrate: %w", err)
+	}
+	if len(res) > 1 {
+		return fmt.Errorf("expected 1 message, got %d", len(res))
+	}
+	return nil
+}
+
+func (d *Device) GetInfo() (Info, error) {
+	if err := d.updateInfo(); err != nil {
+		return Info{}, err
+	}
+	return d.li.info, nil
+}
+
+func (d *Device) updateInfo() error {
+	c, err := netlink.Dial(unix.NETLINK_ROUTE, &netlink.Config{})
+	if err != nil {
+		return fmt.Errorf("couldn't dial netlink socket: %w", err)
+	}
+	defer c.Close()
+
+	ifi := &ifInfoMsg{}
+	ifi.Index = d.index
+	req, err := d.newRequest(unix.RTM_GETLINK, ifi)
+	if err != nil {
+		return fmt.Errorf("couldn't create netlink request: %w", err)
+	}
+
+	res, err := c.Execute(req)
+	if err != nil {
+		return fmt.Errorf("couldn't retrieve link info: %w", err)
+	}
+	if len(res) > 1 {
+		return fmt.Errorf("expected 1 message, got %d", len(res))
+	}
+
+	if err := d.unmarshalBinary(res[0].Data); err != nil {
+		return fmt.Errorf("couldn't decode info: %w", err)
+	}
+	return nil
+}
+
+func (d *Device) newRequest(typ netlink.HeaderType, ifi *ifInfoMsg) (netlink.Message, error) {
+	req := netlink.Message{
+		Header: netlink.Header{
+			Flags: netlink.Request | netlink.Acknowledge,
+			Type:  typ,
+		},
+	}
+	msg := ifi.marshalBinary()
+	req.Data = append(req.Data, msg...)
+	return req, nil
+}
+
+func (ifi *ifInfoMsg) marshalBinary() []byte {
+	buf := make([]byte, unix.SizeofIfInfomsg)
+	buf[0] = ifi.Family
+	buf[1] = 0 // reserved
+	nlenc.PutUint16(buf[2:4], ifi.Type)
+	nlenc.PutInt32(buf[4:8], ifi.Index)
+	nlenc.PutUint32(buf[8:12], ifi.Flags)
+	nlenc.PutUint32(buf[12:16], ifi.Change)
+	return buf
+}
+
+func (ifi *ifInfoMsg) unmarshalBinary(data []byte) error {
+	if len(data) != unix.SizeofIfInfomsg {
+		return fmt.Errorf("data is not a valid ifInfoMsg, expected: %d bytes, got: %d bytes", unix.SizeofIfInfomsg, len(data))
+	}
+	ifi.Family = nlenc.Uint8(data[0:1])
+	ifi.Type = nlenc.Uint16(data[2:4])
+	ifi.Index = nlenc.Int32(data[4:8])
+	ifi.Flags = nlenc.Uint32(data[8:12])
+	ifi.Change = nlenc.Uint32(data[12:16])
+	return nil
+}
+
+func (bt *BitTiming) marshalBinary() []byte {
+	buflen := int(unsafe.Sizeof(BitTiming{}))
+	buf := make([]byte, buflen)
+	nlenc.PutUint32(buf[0:4], bt.Bitrate)
```

Comment:
This can be more safely done by using sizeof for the offsets. I have seen some posts describing that pattern but can't find right now.

Reply:
unsafe.SizeOf is evaluated at compile time, som I'm guessing putting this as a global const would work? 
`const sizeOfBitTiming = int(unsafe.Sizeof(BitTiming{}))`